editor/src/clj/util/http_server.clj

```diff
@@ -295,68 +317,81 @@
+                                                 nil
+                                                 connection)]
+                                (reduce-kv #(doto ^Headers %1 (.add %2 %3)) (.getResponseHeaders exchange) headers)
+                                (.sendResponseHeaders
+                                  exchange
+                                  status
+                                  ;; Response length argument:
+                                  ;; -1 means empty response body (0 bytes)
+                                  ;; 0 is unknown size (chunked)
+                                  ;; positive is exact size
+                                  (if (nil? connection)
+                                    -1
+                                    (if-let [explicit-length (get headers "content-length")]
+                                      (let [n (Long/parseUnsignedLong explicit-length)]
+                                        (if (zero? n) -1 n))
+                                      0)))
+                                (when connection
+                                  (with-open [is (io/input-stream connection)]
+                                    (io/copy is (.getResponseBody exchange)))))
+                              ;; A browser or remote http client can close the
+                              ;; connection before we finished sending the response
+                              ;; headers/body: we only log such exceptions since
+                              ;; they are not an issue that needs to be fixed.
+                              (catch IOException e
+                                (log/warn :exception e))
                               (finally
-                                (when (instance? AutoCloseable connection)
-                                  (.close ^AutoCloseable connection))))))
+                                (when (instance? Closeable connection)
+                                  (.close ^Closeable connection))))))
                         (catch Throwable e
                           (error-reporting/report-exception! e))
                         (finally
                           (.close exchange)))))))))))
     (.start server)
-    (when-not (Boolean/getBoolean "defold.tests")
-      (log/info :msg "Http server running" :local-url (local-url server)))
-    server))
+    (let [server (->ServerWithHandler server handler)]
+      (when-not (Boolean/getBoolean "defold.tests")
+        (log/info :msg "Http server running" :local-url (local-url server)))
+      server)))
+
+(defn handler
+  "Get the server request handler function"
+  [^ServerWithHandler server]
+  (.-handler server))
 
 (defn stop!
-  ^HttpServer
+  ^ServerWithHandler
   ([server] (stop! server 2))
-  ([^HttpServer server delay-seconds]
-   (doto server (.stop delay-seconds))))
+  ([^ServerWithHandler server delay-seconds]
+   (.stop ^HttpServer (.-server server) delay-seconds)
```

Comment:
We could potentially get rid of this type hint if we type-hint the field in the `deftype`.

Reply:
Not this one: field typehints inside deftype hint the values only within the deftype form.